### PR TITLE
fix: Implement urlencoding

### PR DESCRIPTION
**Pull request by bubu11e**
Implement urlencoding for class, labels key and labels value before sending data to warp10.

### CODE REVIEW

File: src/lib/transcompiler.rs [hunk 1]
```diff
@@ -66,21 +69,23 @@ fn format_prometheus(line: &str, now: i64) -> Result<String, Box<dyn Error>> {
...
-                .map(|v| v.replace("\"}", "")) // remove right double quote
-                .map(|v| v.replace(",", "%2C")) // escape
-                .map(|v| v.replace("}", "%7D")) // escape
-                .map(|v| v.replace(r"\\", r"\")) // unescape
-                .map(|v| v.replace("\\\"", "\"")) // unescape
-                .map(|v| v.replace(r"\n", "%0A")) // unescape
+                .split(",")
+                .map(|v| v.replace("=\"", "="))
+                .map(|v| v.replace("\"}", ""))
+                .map(|v| v.replace("\"", ""))
```
Comment (Alkorin) [inline src/lib/transcompiler.rs:81]:
```suggestion
                .map(|v| v.replace("}", "")) // Remove last curly brace
                .map(|v| v.replace("\"", "")) // Remove quotes
```

**Review by Alkorin — APPROVED** [review]

File: Cargo.toml [hunk 2]
```diff
@@ -37,6 +37,7 @@ uuid = { version = "0.8.1", features = ["v4"] }
 slog = { version = "2.5.2", features = ["release_max_level_trace", "max_level_trace"] }
 ctrlc = { version = "3.1.4", features = ["termination"] }
 notify = "4.0.15"
+urlencoding = "1.0.0"
```
Comment (FlorentinDUBOIS) [inline Cargo.toml:40]: The latest version of this package is the 1.3.3

File: src/lib/transcompiler.rs [hunk 3]
```diff
@@ -2,6 +2,9 @@ use std::error::Error;
 
 use time::now_utc;
 
+extern crate urlencoding;
```
Comment (FlorentinDUBOIS) [inline src/lib/transcompiler.rs:5]: The instruction `extern crate` is useless as you import the crate just below using `use`. The only benefit of using `extern crate` is by combining it with `#[macro_use]` at the beginning of the `main.rs` file

File: src/lib/transcompiler.rs [hunk 4]
```diff
@@ -66,29 +69,37 @@ fn format_prometheus(line: &str, now: i64) -> Result<String, Box<dyn Error>> {
...
-                .map(|v| v.replace(",", "%2C")) // escape
-                .map(|v| v.replace("}", "%7D")) // escape
-                .map(|v| v.replace(r"\\", r"\")) // unescape
-                .map(|v| v.replace("\\\"", "\"")) // unescape
-                .map(|v| v.replace(r"\n", "%0A")) // unescape
-                .fold(String::new(), |acc, x| {
-                    // skip invalid values
-                    if !x.contains('=') {
-                        return acc;
+            let mut labels = String::from("");
```
Comment (FlorentinDUBOIS) [inline src/lib/transcompiler.rs:77]:
```suggestion
            let mut labels = String::new();
```

File: src/lib/transcompiler.rs [hunk 5]
```diff
@@ -66,29 +69,37 @@ fn format_prometheus(line: &str, now: i64) -> Result<String, Box<dyn Error>> {
...
+            let mut in_label = false;
+            let mut buffer = String::from("");
```
Comment (FlorentinDUBOIS) [inline src/lib/transcompiler.rs:79]:
```suggestion
            let mut buffer = String::new();
```

File: src/lib/transcompiler.rs [hunk 6]
```diff
@@ -66,29 +69,37 @@ fn format_prometheus(line: &str, now: i64) -> Result<String, Box<dyn Error>> {
...
+            for c in plabels.chars() {
+                if c == '"' {
+                    in_label = !in_label;
+                    continue;
+                }
+
+                if !in_label {
+                    if c == '=' || c == ',' || c == '}' {
+                        labels.push_str(& encode(&buffer));
+                        buffer = String::from("");
```
Comment (FlorentinDUBOIS) [inline src/lib/transcompiler.rs:89]:
```suggestion
                        buffer = String::new();
```

**Review by FlorentinDUBOIS — CHANGES_REQUESTED** [review]
Thanks for the pull request! Could you add some unit tests around the encoding?

**Review by FlorentinDUBOIS — APPROVED** [review]

**Pull request merged**